src/mpid/ch4/shm/xpmem/xpmem_recv.h

```diff
@@ -9,72 +9,316 @@
+                slmt_cts_hdr->sreq_ptr, slmt_cts_hdr->rreq_ptr, slmt_cts_hdr->remote_counter_addr,
+                slmt_cts_hdr->dest_lrank, MPIDIG_REQUEST(rreq, rank));
+
+    /* Receiver sends CTS packet to sender */
+    mpi_errno =
+        MPIDI_SHM_do_ctrl_send(MPIDIG_REQUEST(rreq, rank), comm, MPIDI_SHM_XPMEM_SEND_LMT_CTS,
+                               &ctrl_hdr);
+    MPIR_ERR_CHECK(mpi_errno);
 
-    XPMEM_TRACE("handle_lmt_recv: handle matched rreq %p [source %d, tag %d, context_id 0x%x],"
-                " copy dst %p, src %p, bytes %ld\n", rreq, MPIDIG_REQUEST(rreq, rank),
+    /* Set receive status */
+    MPIR_STATUS_SET_COUNT(rreq->status, recv_data_sz);
+    rreq->status.MPI_SOURCE = MPIDIG_REQUEST(rreq, rank);
+    rreq->status.MPI_TAG = MPIDIG_REQUEST(rreq, tag);
+    dest_buf = MPIDIG_REQUEST(rreq, buffer);
+
+    mpi_errno =
+        MPIDI_XPMEM_seg_regist(src_lrank, src_data_sz, (void *) src_offset, &seg_ptr, &src_buf,
+                               &MPIDI_XPMEM_global.segmaps[src_lrank].segcache_user);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    XPMEM_TRACE("handle_lmt_coop_recv: handle matched rreq %p [source %d, tag %d, context_id 0x%x],"
+                " copy dst %p, src %lx, bytes %ld\n", rreq, MPIDIG_REQUEST(rreq, rank),
                 MPIDIG_REQUEST(rreq, tag), MPIDIG_REQUEST(rreq, context_id),
-                (char *) MPIDIG_REQUEST(rreq, buffer), attached_sbuf, recv_data_sz);
+                (char *) MPIDIG_REQUEST(rreq, buffer), src_offset, recv_data_sz);
+    /* sender and receiver datatypes are both continuous, perform cooperative copy. */
+    mpi_errno =
+        MPIDI_XPMEM_do_lmt_coop_copy(src_buf, recv_data_sz,
+                                     (char *) dest_buf + dt_true_lb,
+                                     &MPIDI_XPMEM_REQUEST(rreq, counter_ptr)->counter, sreq_ptr,
+                                     rreq, &fin_type, &copy_type);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    /* - For receiver:
+     *     case copy_type == LOCAL_FIN:
+     *        case fin_type == COPY_ALL: receiver copies all chunks, complete rreq; free counter obj when receiving FREE_CNT
+     *        case fin_type == COPY_ZERO: receiver copies zero chunk, waits SEND_FIN from sender to complete rreq and free counter obj
+     *        case fin_type == MIXED_COPIED: receiver copies a part of chunks, waits SEND_FIN from sender to complete rreq and free counter obj
+     *     case copy_type == BOTH_FIN:
+     *        case fin_type == COPY_ALL: receiver copies all chunks, complete rreq and free counter obj; send RECV_FIN to sender
+     *        case fin_type == COPY_ZERO: receiver copies zero chunk, complete rreq and free counter obj
+     *        case fin_type == MIXED_COPIED: receiver copies a part of chunks, complete rreq and free counter obj; send RECV_FIN to sender */
+    if ((fin_type == MPIDI_XPMEM_LOCAL_FIN && copy_type == MPIDI_XPMEM_COPY_ALL) ||
```

Comment:
Same suggestion as that in sender side CTS callback. Can you please compress the comments?